### PR TITLE
Restructure Communicator

### PR DESCRIPTION
**Pull request by mrdrogdrog**
### Component/Part
Communicator

### Description
This PR restructures the iframe communicator to remove a lot of common code.

### Steps

- [x] Added implementation
- [x] Added / updated documentation
- [x] I read the [contribution documentation](https://github.com/hedgedoc/react-client/blob/main/CONTRIBUTING.md) and signed-off my commits to accept the DCO.

### CODE REVIEW

File: src/components/editor-page/document-bar/document-info/document-info-line-word-count.tsx [hunk 1]
```diff
@@ -4,38 +4,38 @@
...
+    CommunicationMessageType.ON_WORD_COUNT_CALCULATED,
+    useCallback((values: OnWordCountCalculatedMessage) => setWordCount(values.words), [setWordCount])
+  )
 
-  useEffect(() => {
-    if (rendererReady) {
-      iframeEditorToRendererCommunicator.sendGetWordCount()
-    }
-  }, [iframeEditorToRendererCommunicator, rendererReady])
+  useDoIfRendererReady(
```
Comment (DerMolly) [inline src/components/editor-page/document-bar/document-info/document-info-line-word-count.tsx:34]: maybe `useCallIfRenderIsReady` would be a better name here?

File: src/components/editor-page/renderer-pane/render-iframe.tsx [hunk 2]
```diff
@@ -6,15 +6,27 @@
...
+  SetScrollStateMessage
+} from '../../render-page/window-post-message-communicator/rendering-message'
+import { useEditorToRendererCommunicator } from '../render-context/iframe-editor-to-renderer-communicator-context-provider'
 import { ScrollState } from '../synced-scroll/scroll-props'
 import { useOnIframeLoad } from './hooks/use-on-iframe-load'
 import { ShowOnPropChangeImageLightbox } from './show-on-prop-change-image-lightbox'
 import { setRendererStatus } from '../../../redux/renderer-status/methods'
+import { useEditorReceiveHandler } from '../../render-page/window-post-message-communicator/hooks/use-editor-receive-handler'
+import { useDoIfRendererReady } from '../../render-page/window-post-message-communicator/hooks/use-do-if-renderer-ready'
+import { useRendererReady } from '../../render-page/window-post-message-communicator/hooks/use-renderer-ready'
```
Comment (DerMolly) [inline src/components/editor-page/renderer-pane/render-iframe.tsx:28]: maybe change this to `useRendererIsReady`

File: src/components/render-page/hooks/use-image-click-handler.ts [hunk 3]
```diff
@@ -6,19 +6,23 @@
...
-        src: image.src,
-        alt: image.alt,
-        title: image.title
+      iframeCommunicator.sendMessageToOtherSide({
+        type: CommunicationMessageType.IMAGE_CLICKED,
+        details: {
+          src: image.src,
+          alt: image.alt,
+          title: image.title
+        }
```
Comment (DerMolly) [inline src/components/render-page/hooks/use-image-click-handler.ts:25]: Maybe put the contents of `details` in the top level object as you do everywhere else?

**Review by DerMolly — CHANGES_REQUESTED** [review]
see comments above

File: src/components/editor-page/render-context/iframe-editor-to-renderer-communicator-context-provider.tsx [hunk 4]
```diff
@@ -5,34 +5,29 @@
...
   }
   return communicatorFromContext
 }
 
-export const IframeEditorToRendererCommunicatorContextProvider: React.FC = ({ children }) => {
-  const currentIFrameCommunicator = useMemo<IframeEditorToRendererCommunicator>(
-    () => new IframeEditorToRendererCommunicator(),
-    []
-  )
+export const EditorToRendererCommunicatorContextProvider: React.FC = ({ children }) => {
```
Comment (ErikMichelson) [inline src/components/editor-page/render-context/iframe-editor-to-renderer-communicator-context-provider.tsx:26]: The exported component is called EditorToRendererCommunicatorContextProvider. The filename is different.

File: src/components/render-page/hooks/use-image-click-handler.ts (hunk 3 above)
Comment (mrdrogdrog) [reply src/components/render-page/hooks/use-image-click-handler.ts:25]: Actually this interface is used in CommunicatorImageLightbox :thinking:

File: src/components/editor-page/render-context/editor-to-renderer-communicator-context-provider.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ */
+export const useEditorToRendererCommunicator: () => EditorToRendererCommunicator = () => {
+  const communicatorFromContext = useContext(EditorToRendererCommunicatorContext)
+  if (!communicatorFromContext) {
+    throw new Error('No editor-to-renderer-iframe-communicator received. Did you forget to use the provider component?')
+  }
+  return communicatorFromContext
+}
+
+export const EditorToRendererCommunicatorContextProvider: React.FC = ({ children }) => {
```
Comment (ErikMichelson) [inline src/components/editor-page/render-context/editor-to-renderer-communicator-context-provider.tsx:26]: Missing ESDoc (does not need @return, but should be present)

Comment (mrdrogdrog) [reply src/components/editor-page/render-context/editor-to-renderer-communicator-context-provider.tsx:29]: done

File: src/components/render-page/window-post-message-communicator/window-post-message-communicator.ts [hunk 6]
```diff
@@ -71,8 +91,31 @@ export abstract class IframeCommunicator<SEND, RECEIVE> {
         `Communication isn't enabled. Maybe the other side is not ready?\nMessage was: ${JSON.stringify(message)}`
       )
     }
+    console.debug('[WPMC' + this.generateLogIdentifier() + '] Sent event', message)
     this.messageTarget.postMessage(message, this.targetOrigin)
   }
 
-  protected abstract handleEvent(event: MessageEvent<RECEIVE>): void
+  public setHandler<R extends RECEIVE_TYPE>(eventType: R, handler: Handler<MESSAGES, R>): void {
```
Comment (ErikMichelson) [inline src/components/render-page/window-post-message-communicator/window-post-message-communicator.ts:98]: ESDoc

File: src/components/render-page/window-post-message-communicator/window-post-message-communicator.ts [hunk 7]
```diff
@@ -71,8 +91,31 @@ export abstract class IframeCommunicator<SEND, RECEIVE> {
...
+   * @return {@code true} if the event was processed.
+   */
+  protected handleEvent(event: MessageEvent<PostMessage<RECEIVE_TYPE>>): boolean | undefined {
+    const data = event.data
+
+    const handler = this.handlers[data.type]
+    if (!handler) {
+      return true
+    }
+    console.debug('[WPMC' + this.generateLogIdentifier() + '] Received event ', data)
```
Comment (ErikMichelson) [inline src/components/render-page/window-post-message-communicator/window-post-message-communicator.ts:117]: missing space between WPMC and log identifier

**Review by ErikMichelson — APPROVED** [review]

File: src/components/render-page/hooks/use-image-click-handler.ts (hunk 3 above)
Comment (DerMolly) [reply src/components/render-page/hooks/use-image-click-handler.ts:25]: Maybe we should explicity use that type here then?

File: src/components/editor-page/editor-page.tsx [hunk 8]
```diff
@@ -79,10 +87,12 @@ export const EditorPage: React.FC = () => {
 
   const setRendererToScrollSource = useCallback(() => {
     scrollSource.current = ScrollSource.RENDERER
+    console.debug('[EditorPage] Make renderer scroll source')
```
Comment (DerMolly) [inline src/components/editor-page/editor-page.tsx:90]: Should these be still here?

File: src/components/editor-page/render-context/editor-to-renderer-communicator-context-provider.tsx [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import React, { createContext, useContext, useMemo } from 'react'
+import { EditorToRendererCommunicator } from '../../render-page/window-post-message-communicator/editor-to-renderer-communicator'
+
+const EditorToRendererCommunicatorContext = createContext<EditorToRendererCommunicator | undefined>(undefined)
+
+/**
+ * Provides the {@link EditorToRendererCommunicator editor to renderer iframe communicator} that is set by a {@link EditorToRendererCommunicatorContextProvider context provider}.
+ *
+ * @return the received communicator
+ * @throws Error if no communicator was received
```
Comment (DerMolly) [inline src/components/editor-page/render-context/editor-to-renderer-communicator-context-provider.tsx:16]:
```suggestion
 * @throws {Error} if no communicator was received
```

I'm not sure if we do that elsewhere in this way, but the type should be in curly brackets see [here](https://esdoc.org/manual/tags.html#-return)

File: src/components/editor-page/render-context/renderer-to-editor-communicator-context-provider.tsx [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+import { RendererToEditorCommunicator } from '../../render-page/window-post-message-communicator/renderer-to-editor-communicator'
+import { CommunicationMessageType } from '../../render-page/window-post-message-communicator/rendering-message'
+
+const RendererToEditorCommunicatorContext = createContext<RendererToEditorCommunicator | undefined>(undefined)
+
+/**
+ * Provides the {@link RendererToEditorCommunicator renderer to editor iframe communicator} that is set by a {@link RendererToEditorCommunicatorContextProvider context provider}.
+ *
+ * @return the received communicator
+ * @throws Error if no communicator was received
```
Comment (DerMolly) [inline src/components/editor-page/render-context/renderer-to-editor-communicator-context-provider.tsx:19]: Same here

File: src/components/render-page/window-post-message-communicator/window-post-message-communicator.ts [hunk 11]
```diff
@@ -9,19 +9,39 @@
...
+export type HandlerMap<MESSAGES, MESSAGE_TYPE extends string> = Partial<{
+  [key in MESSAGE_TYPE]: Handler<MESSAGES, MESSAGE_TYPE>
+}>
+
+export interface PostMessage<MESSAGE_TYPE extends string> {
+  type: MESSAGE_TYPE
+}
+
 /**
  * Base class for communication between renderer and editor.
```
Comment (DerMolly) [inline src/components/render-page/window-post-message-communicator/window-post-message-communicator.ts:25]: Shouldn't this add the parameters here?

**Review by DerMolly — CHANGES_REQUESTED** [review]

File: src/components/editor-page/editor-page.tsx (hunk 8 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-page.tsx:90]: Yes, @mrdrogdrog and I discussed that and decided to leave that in and use a proper logging framework later on (see #1511).

File: src/components/render-page/window-post-message-communicator/window-post-message-communicator.ts (hunk 11 above)
Comment (mrdrogdrog) [reply src/components/render-page/window-post-message-communicator/window-post-message-communicator.ts:25]: Why should it? The most basic message you have has no parameters. If you take a look into render-messages.ts then you'll find the interfaces for the specific messages with parameters that extend this interface.

File: src/components/editor-page/editor-page.tsx (hunk 8 above)
Comment (mrdrogdrog) [reply src/components/editor-page/editor-page.tsx:90]: Won't change this here

File: src/components/render-page/window-post-message-communicator/window-post-message-communicator.ts (hunk 11 above)
Comment (DerMolly) [reply src/components/render-page/window-post-message-communicator/window-post-message-communicator.ts:25]: Fair. I missed the `abstract`

**Review by DerMolly — APPROVED** [review]

**Pull request merged**